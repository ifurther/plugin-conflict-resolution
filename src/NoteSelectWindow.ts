import JoplinData from "api/JoplinData";
import JoplinViewsDialogs from "api/JoplinViewsDialogs";

export class NoteSelectWindow {
    private joplinDialogs : JoplinViewsDialogs;
    private joplinData : JoplinData;
    private joplinInstallDir : string;
    private fileSystem;

    constructor(joplinDialogs : JoplinViewsDialogs, joplinData : JoplinData, fileSystem, joplinInstallDir : string) {
        this.joplinDialogs = joplinDialogs;
        this.joplinData = joplinData;
        this.fileSystem = fileSystem;
        this.joplinInstallDir = joplinInstallDir;
    }

    private handle : string = "";

    public async init(id : string) {
        this.handle = await this.joplinDialogs.create(id);
        await this.joplinDialogs.addScript(this.handle, "./ui/NoteSelectWindow/index.js");
        await this.joplinDialogs.addScript(this.handle, "./ui/NoteSelectWindow/index.css");
    }

    public async openDialog() : Promise<string> {

        const notesList = await this.getNotes();
        const escapedJSON = JSON.stringify(notesList).replace(/"/g, '&quot;');

        const htmlContents = await this.fileSystem.readFile(this.joplinInstallDir + "/ui/NoteSelectWindow/index.html");
        await this.joplinDialogs.setHtml(this.handle, `
            <input id="notesList" type="hidden" value="${escapedJSON}" />
            ${htmlContents}
        `);

        const result = await this.joplinDialogs.open(this.handle);
        
        if(result.id !== 'ok') {
            return "";
        }

        if(result === undefined || result.formData === undefined || result.formData.noteSelectForm === undefined || result.formData.noteSelectForm.noteSelect === undefined)
            throw new Error("No note was selected!");
        
        return result.formData.noteSelectForm.noteSelect;
    }

    private async getNotes() {
        const notes = await this.joplinData.get(['notes'], { fields: ['parent_id', 'title', 'id'], order_by: 'updated_time', order_dir: 'DESC' });
 
        // TODO: Handle the hasMore attribute!

        let notesList = [];
        for(const note of notes.items) {
            
            let currentParent = note.parent_id;
            let curFolderTitle = "";
            while(currentParent !== "") {
                if(curFolderTitle !== "")
                    curFolderTitle = " / " + curFolderTitle;
                let curFolder = await this.joplinData.get(['folders', currentParent], { fields: ['title', 'parent_id'] });
                curFolderTitle = curFolder.title + curFolderTitle;
                currentParent = curFolder.parent_id;
            }
            
            notesList.push({
                id: note.id,
                title: note.title,
                folderName: curFolderTitle
            });
        }

        return notesList;
    }
}