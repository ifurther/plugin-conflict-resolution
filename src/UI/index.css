.conflictRes-Editor {
    width: 700px;
    height: 500px;
}