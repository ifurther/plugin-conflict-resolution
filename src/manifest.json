{
	"manifest_version": 1,
	"id": "joplin-plugin-conflict-resolution",
	"app_min_version": "2.0.9",
	"version": "1.2.0",
	"name": "Conflict Resolution",
	"description": "A plugin that makes process of resolving conflicts easier.",
	"author": "Ahmed Alwasifey",
	"homepage_url": "https://github.com/joplin/plugin-conflict-resolution",
	"repository_url": "https://github.com/joplin/plugin-conflict-resolution",
	"keywords": []
}