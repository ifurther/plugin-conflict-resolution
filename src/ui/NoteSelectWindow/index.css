body {
    width: 500px;
}

#autoComplete_wrapper {
    height: 500px;
    /*display: block;
    position: absolute;
    z-index: 100;*/
}