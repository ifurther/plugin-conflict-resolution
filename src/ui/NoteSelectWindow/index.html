<div id="autoComplete_wrapper">
    <input id="autoComplete">
</div>