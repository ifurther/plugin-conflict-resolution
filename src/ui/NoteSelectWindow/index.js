setTimeout(init, 5, 10);

function init() {
    const notesList = JSON.parse(document.getElementById("notesList").value);

    let notesObject = {};
    let orderedNotebooks = [];

    notesList.forEach(note => {
        if(notesObject[note.folderName] === undefined) {
            notesObject[note.folderName] = [];
            orderedNotebooks.push(note.folderName);
        }

        notesObject[note.folderName].push({
            id: note.id,
            title: note.title == "" ? "Untitled" : note.title
        });
    });

    const selectItem = document.getElementById("noteSelect");

    for(const folder of orderedNotebooks) {
        const groupHtml = document.createElement("optgroup");
        groupHtml.setAttribute("label", folder);
        notesObject[folder].forEach(note => {
            const noteHtml = document.createElement("option");
            noteHtml.setAttribute("value", note.id);
            noteHtml.innerText = note.title;
            groupHtml.appendChild(noteHtml);
        });
        selectItem.appendChild(groupHtml);
    }
}