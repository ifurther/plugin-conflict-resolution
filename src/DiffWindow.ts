import JoplinData from "api/JoplinData";
import JoplinViewsDialogs from "api/JoplinViewsDialogs";
import { NoteSelectWindow } from "./NoteSelectWindow";

export class DiffWindow {
    
    private joplinDialogs : JoplinViewsDialogs;
    private joplinData : JoplinData;
    private joplinInstallDir : string;
    private fileSystem;
    private noteSelectWindow : NoteSelectWindow;

    constructor(joplinDialogs : JoplinViewsDialogs, joplinData : JoplinData, fileSystem, joplinInstallDir : string) {
        this.joplinDialogs = joplinDialogs;
        this.joplinData = joplinData;
        this.joplinInstallDir = joplinInstallDir;
        this.fileSystem = fileSystem;
    }


    private handle: string = "";
    
    public async Init(id : string) {
        this.handle = await this.joplinDialogs.create(id);
		await this.joplinDialogs.addScript(this.handle, './lib/codemirror/lib/codemirror.js');
		await this.joplinDialogs.addScript(this.handle, './lib/codemirror/lib/codemirror.css');
		await this.joplinDialogs.addScript(this.handle, './lib/diff_match_patch/diff_match_patch.js');
		await this.joplinDialogs.addScript(this.handle, './lib/codemirror/addon/merge/merge.css');
		await this.joplinDialogs.addScript(this.handle, './ui/DiffWindow/index.js');
		await this.joplinDialogs.addScript(this.handle, './ui/DiffWindow/index.css');
		await this.joplinDialogs.setButtons(this.handle, [
			{
				id: 'submit',
				title: 'Save'
			},
			{
				id: 'cancel',
				title: 'Cancel'
			}
		]);

        this.noteSelectWindow = new NoteSelectWindow(this.joplinDialogs, this.joplinData, this.fileSystem, this.joplinInstallDir);
        await this.noteSelectWindow.init("dialog-note-select");
    }

    public async OpenWindow(noteId: string, compareWithId : string = "") {
        const localNote = await this.joplinData.get(['notes', noteId], {
            fields: ['is_conflict', 'conflict_original_id', 'body', 'title']
        });
    
        if(localNote.is_conflict === 0) {
            throw new Error('This is not a conflict note.');
        }
    
        if(localNote.conflict_original_id === "" && compareWithId === "") {
            compareWithId = await this.noteSelectWindow.openDialog();
        }
    
        const remoteId = (compareWithId == "" ? localNote.conflict_original_id : compareWithId);

        const remoteNote = await this.joplinData.get(['notes', remoteId], {
            fields: ['body', 'title']
        });
    
        const remoteNoteContent = remoteNote.body.replace(/"/g, '&quot;');
        const localNoteContent = localNote.body.replace(/"/g, '&quot;');
        const remoteNoteTitle = remoteNote.title.replace(/"/g, '&quot;');
        const localNoteTitle = localNote.title.replace(/"/g, '&quot;');
    
        const htmlContents = await this.fileSystem.readFile(this.joplinInstallDir + '/ui/DiffWindow/index.html');

        // These inputs are a simple hack in order to pass data into the WebView.
        await this.joplinDialogs.setHtml(this.handle, `
            <input id="pluginInstallDir" type="hidden" value="${this.joplinInstallDir}"/> 
            <input id="remoteNote" type="hidden" value="${remoteNoteContent}"/> 
            <input id="curNote" type="hidden" value="${localNoteContent}"/> 
            <input id="remoteTitle" type="hidden" value="${remoteNoteTitle}"/> 
            <input id="curTitle" type="hidden" value="${localNoteTitle}"/> 
            ${htmlContents}
        `);
    
        const response = await this.joplinDialogs.open(this.handle);
        return response;
    }
}