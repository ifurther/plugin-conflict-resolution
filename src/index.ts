import joplin from 'api';
import {MenuItemLocation} from 'api/types';
import {DiffWindow} from './DiffWindow';
import {NoteSelectWindow} from './NoteSelectWindow';

let diffWindow : DiffWindow;
let noteSelectWindow : NoteSelectWindow;

/**
 * Opens a Diff Window to resolve a conflict.
 * @param {string[]} noteIds Should contain the note to resolve the conflict for.
 */
async function openDiffWindow(noteIds : string[]) {
    if (noteIds.length !== 1) {
    // Comparing multiple notes not yet supported.
        return;
    }

    const noteId = noteIds[0];
    const localNote = await joplin.data.get(['notes', noteId], {
        fields: ['is_conflict', 'conflict_original_id'],
    });

    if (localNote.is_conflict === 0) {
        return await joplin.views.dialogs.showMessageBox('This is not a conflict note.');
    }

    let compareWithId = localNote.conflict_original_id;
    if (compareWithId === '') {
        compareWithId = await noteSelectWindow.openDialog();
    }

    if (compareWithId === '') {
    // No note was found to compare to and user didn't select a note. Cancel button must have been pressed.
        return;
    }

    try {
        const resp = await diffWindow.openWindow(noteId, compareWithId);
        if (resp === null) {
            // If it's null, the Cancel button was pressed.
            return;
        }

        // The note in the normal folder is actually the note that came from the remote server, AKA the one in compareWithId.
        // So, we must replace the contents of that note. The note at `noteId` exists in conflicts folder and will be removed.
        await joplin.data.put(['notes', compareWithId], null, {
            body: resp.NoteContents,
            title: resp.NoteTitle,
        });

        await joplin.data.delete(['notes', noteId]);

        await joplin.views.dialogs.showMessageBox('Successfully resolved conflict!');
    } catch (ex) {
        joplin.views.dialogs.showMessageBox(ex.message);
    }
}

joplin.plugins.register({
    onStart: async function() {
        diffWindow = new DiffWindow(joplin.views.dialogs, joplin.data, joplin.require('fs-extra'), await joplin.plugins.installationDir());
        await diffWindow.init('conflictResolution-diff-dialog');

        noteSelectWindow = new NoteSelectWindow(joplin.views.dialogs, joplin.data, joplin.require('fs-extra'), await joplin.plugins.installationDir());
        await noteSelectWindow.init('conflictResolution-noteSelect-dialog');

        await joplin.commands.register({
            name: 'resolveConflictsCommand',
            label: 'Resolve Conflict',
            enabledCondition: 'oneNoteSelected && inConflictFolder',
            execute: openDiffWindow,
        });

        joplin.views.menuItems.create('resolveConflictsButton', 'resolveConflictsCommand', MenuItemLocation.NoteListContextMenu);

        console.info('Conflict Resolution Plugin loaded!');
    },
});
