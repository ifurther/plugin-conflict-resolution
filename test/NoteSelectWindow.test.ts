const JoplinViewsDialogs = require("./../api/JoplinViewsDialogs.d.ts");
const JoplinData = require("./../api/JoplinData.d.ts");
import { NoteSelectWindow } from "../src/NoteSelectWindow";

/* JOPLIN API MOCK */
const mockedDialogApi = {
    create: jest.fn(),
    addScript: jest.fn(),
    setHtml: jest.fn(),
    open: jest.fn()
}
jest.mock("./../api/JoplinViewsDialogs.d.ts", () => {
    return jest.fn().mockImplementation(() => {
        return {
            create: mockedDialogApi.create,
            addScript: mockedDialogApi.addScript,
            setHtml: mockedDialogApi.setHtml,
            open: mockedDialogApi.open
        };
    });
});

const mockedDataApi = {
    get: jest.fn()
}
jest.mock("./../api/JoplinData.d.ts", () => {
    return jest.fn().mockImplementation(() => {
        return {
            get: mockedDataApi.get
        };
    });
});

const fs = require('fs');
jest.mock('fs');

describe('NoteSelectWindow', () => {
    
    const mockedJoplinDialogs = new JoplinViewsDialogs();
    const mockedJoplinData = new JoplinData();

    mockedDialogApi.create.mockResolvedValue("test-dialog-id");
    fs.readFile.mockResolvedValue("<form></form>");
    mockedDialogApi.open.mockResolvedValue({
        id: 'ok',
        formData: {
            noteSelectForm: {
                noteSelect: '6d8c895238114637b63ed54149a4dd84'
            }
        }
    });
    mockedDataApi.get.mockImplementation((arr, obj) => {
        if(arr[0] === 'folders') {
            if(arr[1] === "8c0d6c64f37f409ba4edf517bab861b5") {
                return {
                    title: "Test Folder 1",
                    parent_id: ""
                };
            } else if(arr[1] === "673f0783febc4ead8187727781ab3b32") {
                return {
                    title: "Test Folder 2",
                    parent_id: "8c0d6c64f37f409ba4edf517bab861b5"
                };
            } else if(arr[1] === "deab7b3f223a4a1daf26cea675beb674") {
                return {
                    title: "Test Folder 3",
                    parent_id: "673f0783febc4ead8187727781ab3b32"
                };
            } else if(arr[1] === "e3f8e3c80a0f4a029b35bcd82ef46a06") {
                return {
                    title: "Test Folder 4",
                    parent_id: ""
                };
            }
        } else if(arr[0] == 'notes') {
            if(obj.page && obj.page == 2) {
                return {
                    items:[
                        {
                            "parent_id":"8c0d6c64f37f409ba4edf517bab861b5",
                            "title":"gasserrrrrr",
                            "id":"3667da1054814e3c90d755d26a685a37"
                        },
                        {
                            "parent_id":"673f0783febc4ead8187727781ab3b32",
                            "title":"dg",
                            "id":"f57154854b4c424789bc47cfe2dee85e"
                        },
                        {
                            "parent_id":"673f0783febc4ead8187727781ab3b32",
                            "title":"rwerw",
                            "id":"6355692b5ed2420cbcdfa3b717f9bc8e"
                        },
                        {
                            "parent_id":"8c0d6c64f37f409ba4edf517bab861b5",
                            "title":"sdfsd",
                            "id":"6965f828d1ae4d40acf0a86caa3bec31"
                        },
                        {
                            "parent_id":"8c0d6c64f37f409ba4edf517bab861b5",
                            "title":"sdf",
                            "id":"c00715506623400fab95f62aadf1797a"
                        },
                        {
                            "parent_id":"8c0d6c64f37f409ba4edf517bab861b5",
                            "title":"sd",
                            "id":"fafa879c78aa4977b92b63bbb5859ee8"
                        },
                        {
                            "parent_id":"8c0d6c64f37f409ba4edf517bab861b5",
                            "title":"",
                            "id":"21e6b6eb59fe4284bbdc00cd571851c8"
                        },
                        {
                            "parent_id":"8c0d6c64f37f409ba4edf517bab861b5",
                            "title":"fgerg",
                            "id":"af3d3732f3eb4cb79f334077f73b26da"
                        },
                        {
                            "parent_id":"8c0d6c64f37f409ba4edf517bab861b5",
                            "title":"g",
                            "id":"a7e4bda6fc54415082260a3d4270c568"
                        }
                    ],
                    has_more:true
                };
            } else {
                return {
                    items: [{
                        "parent_id": "8c0d6c64f37f409ba4edf517bab861b5",
                        "title": "g",
                        "id": "f78f5b7f72634a1190af2f17b1e8d96c"
                    },
                    {
                        "parent_id": "8c0d6c64f37f409ba4edf517bab861b5",
                        "title": "",
                        "id": "c4e21d6016584e27b3f30831994a2a6b"
                    },
                    {
                        "parent_id": "8c0d6c64f37f409ba4edf517bab861b5",
                        "title": "dg",
                        "id": "15f338dc3664487189fee5af0750d0d1"
                    },
                    {
                        "parent_id": "8c0d6c64f37f409ba4edf517bab861b5",
                        "title": "Ahmed",
                        "id": "b0f025664fa84645a3eaccb99209e0b8"
                    },
                    {
                        "parent_id": "deab7b3f223a4a1daf26cea675beb674",
                        "title": "Gasser Note",
                        "id": "a5548d09595b4baea839cd7d40ee4aa1"
                    },
                    {
                        "parent_id": "deab7b3f223a4a1daf26cea675beb674",
                        "title": "Anas Note",
                        "id": "8dc676bddd2b43cbbab197e426722ad9"
                    },
                    {
                        "parent_id": "e3f8e3c80a0f4a029b35bcd82ef46a06",
                        "title": "Conflict resolution plugin",
                        "id": "bd1a227764a944bab6eff29eb116e6a8"
                    },
                    {
                        "parent_id": "e3f8e3c80a0f4a029b35bcd82ef46a06",
                        "title": "Conflict resolution plugin",
                        "id": "6d8c895238114637b63ed54149a4dd84"
                    },
                    {
                        "parent_id": "deab7b3f223a4a1daf26cea675beb674",
                        "title": "Ahmed Note",
                        "id": "7fd64520f20548fb94356b032751e72f"
                    },
                    {
                        "parent_id": "e3f8e3c80a0f4a029b35bcd82ef46a06",
                        "title": "Test text is here.",
                        "id": "73e8d938c4d84951915278a54e0eacaf"
                    },
                    {
                        "parent_id": "e3f8e3c80a0f4a029b35bcd82ef46a06",
                        "title": "",
                        "id": "4cac336dc7ce4e449442ed535b93b085"
                    },
                    {
                        "parent_id": "e3f8e3c80a0f4a029b35bcd82ef46a06",
                        "title": "sdfsd",
                        "id": "2432b08c86b84fd9983bee4c42699ca8"
                    },
                    {
                        "parent_id": "e3f8e3c80a0f4a029b35bcd82ef46a06",
                        "title": "dfgdg",
                        "id": "9ad879e68da34a5c8420f45e77c096b6"
                    },
                    {
                        "parent_id": "e3f8e3c80a0f4a029b35bcd82ef46a06",
                        "title": "dsfsdf",
                        "id": "945d9edeba9a48ba90958af0f3c4c66b"
                    }],
                    has_more: false
                };
            }
        }
    });


    beforeEach(() => {
        jest.clearAllMocks();
    });

    it("should test get notes function", async () => {
        const dialog = new NoteSelectWindow(mockedJoplinDialogs, mockedJoplinData, fs, "D:");
        const res = await dialog.getNotes();
        expect(res).toMatchSnapshot("getNotesSnapshot");
    });

    it("should setup the dialog properly", async () => {        
        const dlg = new NoteSelectWindow(mockedJoplinDialogs, mockedJoplinData, fs, "D:");
        
        await dlg.init("abc123");
        expect(mockedDialogApi.create).toBeCalledWith("abc123");
        expect(mockedDialogApi.addScript).toBeCalledTimes(2);
        expect(mockedDialogApi.addScript).toBeCalledWith("test-dialog-id", "./ui/NoteSelectWindow/index.js")
        expect(mockedDialogApi.addScript).toBeCalledWith("test-dialog-id", "./ui/NoteSelectWindow/index.css")
    });

    it("should test the dialog having correct html", async () => {
        const dlg = new NoteSelectWindow(mockedJoplinDialogs, mockedJoplinData, fs, "D:");
        dlg.init("abc123");
        await dlg.openDialog();
        expect(mockedDialogApi.setHtml).toBeCalledTimes(1);
        expect(mockedDialogApi.setHtml.mock.calls[0][0]).toBe("test-dialog-id");
        expect(mockedDialogApi.setHtml.mock.calls[0][1]).toMatchSnapshot("setHtmlFirstSnapshot");
    });

    it("should test ok result", async() => {
        const dlg = new NoteSelectWindow(mockedJoplinDialogs, mockedJoplinData, fs, "D:");
        dlg.init("abc123");
        const res = await dlg.openDialog();
        expect(res).toBe("6d8c895238114637b63ed54149a4dd84");  // This is the note id that is returned by the mocked .open() function.
    });

    it("should raise an error when response doesn't have any note selected", async () => {
        const dlg = new NoteSelectWindow(mockedJoplinDialogs, mockedJoplinData, fs, "D:");
        dlg.init("abc123");
        mockedDialogApi.open.mockResolvedValueOnce({
            id: "ok",
            result: {
                formData: {
                    noteSelectForm: {
                        noteSelect: undefined
                    }
                }
            }
        });

        expect.assertions(1);
        try {
            await dlg.openDialog();
        } catch(e) {
            expect(e).toMatchSnapshot("noNoteSelectedError");
        }
    });

    it("should return empty string if user hits cancel", async () => {
        const dlg = new NoteSelectWindow(mockedJoplinDialogs, mockedJoplinData, fs, "D:");
        dlg.init("abc123");
        mockedDialogApi.open.mockResolvedValueOnce({
            id: "cancel"
        });
        const res = await dlg.openDialog();
        expect(res).toBe("");
    });
});
