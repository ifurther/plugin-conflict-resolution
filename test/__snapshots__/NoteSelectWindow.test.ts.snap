// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteSelectWindow should raise an error when response doesn't have any note selected: noNoteSelectedError 1`] = `[Error: No note was selected!]`;

exports[`NoteSelectWindow should test get notes function: getNotesSnapshot 1`] = `
Array [
  Object {
    "folderName": "Test Folder 1",
    "id": "f78f5b7f72634a1190af2f17b1e8d96c",
    "title": "g",
  },
  Object {
    "folderName": "Test Folder 1",
    "id": "c4e21d6016584e27b3f30831994a2a6b",
    "title": "",
  },
  Object {
    "folderName": "Test Folder 1",
    "id": "15f338dc3664487189fee5af0750d0d1",
    "title": "dg",
  },
  Object {
    "folderName": "Test Folder 1",
    "id": "b0f025664fa84645a3eaccb99209e0b8",
    "title": "Ahmed",
  },
  Object {
    "folderName": "Test Folder 1 / Test Folder 2 / Test Folder 3",
    "id": "a5548d09595b4baea839cd7d40ee4aa1",
    "title": "Gasser Note",
  },
  Object {
    "folderName": "Test Folder 1 / Test Folder 2 / Test Folder 3",
    "id": "8dc676bddd2b43cbbab197e426722ad9",
    "title": "Anas Note",
  },
  Object {
    "folderName": "Test Folder 4",
    "id": "bd1a227764a944bab6eff29eb116e6a8",
    "title": "Conflict resolution plugin",
  },
  Object {
    "folderName": "Test Folder 4",
    "id": "6d8c895238114637b63ed54149a4dd84",
    "title": "Conflict resolution plugin",
  },
  Object {
    "folderName": "Test Folder 1 / Test Folder 2 / Test Folder 3",
    "id": "7fd64520f20548fb94356b032751e72f",
    "title": "Ahmed Note",
  },
  Object {
    "folderName": "Test Folder 4",
    "id": "73e8d938c4d84951915278a54e0eacaf",
    "title": "Test text is here.",
  },
  Object {
    "folderName": "Test Folder 4",
    "id": "4cac336dc7ce4e449442ed535b93b085",
    "title": "",
  },
  Object {
    "folderName": "Test Folder 4",
    "id": "2432b08c86b84fd9983bee4c42699ca8",
    "title": "sdfsd",
  },
  Object {
    "folderName": "Test Folder 4",
    "id": "9ad879e68da34a5c8420f45e77c096b6",
    "title": "dfgdg",
  },
  Object {
    "folderName": "Test Folder 4",
    "id": "945d9edeba9a48ba90958af0f3c4c66b",
    "title": "dsfsdf",
  },
]
`;

exports[`NoteSelectWindow should test the dialog having correct html: setHtmlFirstSnapshot 1`] = `
"
            <input id=\\"notesList\\" type=\\"hidden\\" value=\\"[{&quot;id&quot;:&quot;f78f5b7f72634a1190af2f17b1e8d96c&quot;,&quot;title&quot;:&quot;g&quot;,&quot;folderName&quot;:&quot;Test Folder 1&quot;},{&quot;id&quot;:&quot;c4e21d6016584e27b3f30831994a2a6b&quot;,&quot;title&quot;:&quot;&quot;,&quot;folderName&quot;:&quot;Test Folder 1&quot;},{&quot;id&quot;:&quot;15f338dc3664487189fee5af0750d0d1&quot;,&quot;title&quot;:&quot;dg&quot;,&quot;folderName&quot;:&quot;Test Folder 1&quot;},{&quot;id&quot;:&quot;b0f025664fa84645a3eaccb99209e0b8&quot;,&quot;title&quot;:&quot;Ahmed&quot;,&quot;folderName&quot;:&quot;Test Folder 1&quot;},{&quot;id&quot;:&quot;a5548d09595b4baea839cd7d40ee4aa1&quot;,&quot;title&quot;:&quot;Gasser Note&quot;,&quot;folderName&quot;:&quot;Test Folder 1 / Test Folder 2 / Test Folder 3&quot;},{&quot;id&quot;:&quot;8dc676bddd2b43cbbab197e426722ad9&quot;,&quot;title&quot;:&quot;Anas Note&quot;,&quot;folderName&quot;:&quot;Test Folder 1 / Test Folder 2 / Test Folder 3&quot;},{&quot;id&quot;:&quot;bd1a227764a944bab6eff29eb116e6a8&quot;,&quot;title&quot;:&quot;Conflict resolution plugin&quot;,&quot;folderName&quot;:&quot;Test Folder 4&quot;},{&quot;id&quot;:&quot;6d8c895238114637b63ed54149a4dd84&quot;,&quot;title&quot;:&quot;Conflict resolution plugin&quot;,&quot;folderName&quot;:&quot;Test Folder 4&quot;},{&quot;id&quot;:&quot;7fd64520f20548fb94356b032751e72f&quot;,&quot;title&quot;:&quot;Ahmed Note&quot;,&quot;folderName&quot;:&quot;Test Folder 1 / Test Folder 2 / Test Folder 3&quot;},{&quot;id&quot;:&quot;73e8d938c4d84951915278a54e0eacaf&quot;,&quot;title&quot;:&quot;Test text is here.&quot;,&quot;folderName&quot;:&quot;Test Folder 4&quot;},{&quot;id&quot;:&quot;4cac336dc7ce4e449442ed535b93b085&quot;,&quot;title&quot;:&quot;&quot;,&quot;folderName&quot;:&quot;Test Folder 4&quot;},{&quot;id&quot;:&quot;2432b08c86b84fd9983bee4c42699ca8&quot;,&quot;title&quot;:&quot;sdfsd&quot;,&quot;folderName&quot;:&quot;Test Folder 4&quot;},{&quot;id&quot;:&quot;9ad879e68da34a5c8420f45e77c096b6&quot;,&quot;title&quot;:&quot;dfgdg&quot;,&quot;folderName&quot;:&quot;Test Folder 4&quot;},{&quot;id&quot;:&quot;945d9edeba9a48ba90958af0f3c4c66b&quot;,&quot;title&quot;:&quot;dsfsdf&quot;,&quot;folderName&quot;:&quot;Test Folder 4&quot;}]\\" />
            <form></form>
        "
`;
